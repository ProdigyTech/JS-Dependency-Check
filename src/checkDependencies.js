import axios from "axios";
import semverGte from "semver/functions/gte.js";
import diff from "semver/functions/diff.js";

const NPM_REGISTRY_URL = "https://registry.npmjs.org";
const NPM_PACKAGE_URL = "https://www.npmjs.com/package";
const whitelistedDependencies = process.env.DEP_CHECK_WHITELIST || [];

const filterDependencies = (whiteList, dep) => {
  return dep.filter((d) => {
    return !whiteList.includes(d.package);
  });
};

export const checkDependencies = async ({
  peerDependencies = [],
  devDependencies = [],
  dependencies = [],
}) => {

  const failedLookupResult = [];
  const whiteList =
    whitelistedDependencies.length > 0
      ? whitelistedDependencies.split(",")
      : [];

  // successfulLookups, failedLookups;

  const {
    successfulLookups: peerDependenciesResult,
    failedLookups: failedPeerDependencies,
  } = await processDependencies(peerDependencies, whiteList);

  const {
    successfulLookups: devDependenciesResult,
    failedLookups: failedDevDependencies,
  } = await processDependencies(devDependencies, whiteList);

  const {
    successfulLookups: dependenciesResult,
    failedLookups: failedDependencies,
  } = await processDependencies(dependencies, whiteList);

  failedLookupResult.push(
    ...failedDependencies,
    ...failedDevDependencies,
    ...failedPeerDependencies
  );

  return {
    peerDependenciesResult,
    devDependenciesResult,
    dependenciesResult,
    failedLookupResult,
  };
};

const processDependencies = async (dep, whiteList) => {
  try {
    const failedLookups = [];
    const filteredDeps = filterDependencies(whiteList, dep);
    const processedData = await Promise.all(
      filteredDeps.map(async (current) => {
        const data = await checkDependencyInNPMRegistry({
          package: current.package,
        });
        const report = await generateReport(data, current);
        return report;
      })
    );
    const successfulLookups = processedData.filter((f) => {
      if (f.package.error) {
        failedLookups.push(f);
        return false;
      } else {
        return true;
      }
    });
    return { successfulLookups, failedLookups };
  } catch (e) {
    console.error(e);
    process.exit(1);
  }
};

const checkDependencyInNPMRegistry = async ({ package: jsPackage }) => {
      const url = `${NPM_REGISTRY_URL}/${jsPackage}`;
  try {
    const { data } = await axios.get(url);
    const { time } = data;
    const tags = data["dist-tags"];
    return { versionTimeline: time, tags };
  } catch (e) {
    console.error(
      `There was an issue searching the registry for ${jsPackage}, skipping...`
    );
    return {
      versionTimeline: {},
      tags: {},
      error: true,
      name: jsPackage,
      url,
      stackTrace: e,
    };
  }
};

const generateVersionObject = ({
  name,
  versionTimeline,
  latest,
  definedVersion,
  error = false,
  currentPackage,
  stackTrace,
}) => {
  if (error)
    return {
      package: {
        error,
        name: currentPackage.package,
        version: currentPackage.version,
        stackTrace: stackTrace,
      },
    };

  return {
    package: {
      name,
      registry_url: `${NPM_REGISTRY_URL}/${name}`,
      npm_url: `${NPM_PACKAGE_URL}/${name}`,
      latest: {
        version: latest || definedVersion,
        releaseDate: versionTimeline[latest] || versionTimeline[definedVersion],
      },
      current: {
        version: definedVersion,
        releaseDate: versionTimeline[definedVersion],
      },
      upgradeType: diff(definedVersion, latest || definedVersion) || "N/A",
      error,
    },
  };
};

const generateReport = async (
  { versionTimeline, tags, error = false, stackTrace },
  currentPackage
) => {
  return new Promise((resolve, reject) => {
    try {
      if (error) {
        return resolve(
          generateVersionObject({
            error,
            currentPackage,
            stackTrace,
          })
        );
      }

      const getDefinedVersion = () => {
        if (Number.isNaN(Number.parseFloat(currentPackage.version))) {
          const v = currentPackage.version.split("");
          const [throwAway, ...rest] = v;
          return rest.join("");
        } else {
          return currentPackage.version;
        }
      };

      const definedVersion = getDefinedVersion();

      const { latest } = tags;
      let versionInfo = {};

      if (!semverGte(definedVersion, latest)) {
        versionInfo = generateVersionObject({
          name: currentPackage.package,
          versionTimeline,
          latest,
          definedVersion,
        });
      } else {
        versionInfo = generateVersionObject({
          name: currentPackage.package,
          versionTimeline,
          definedVersion,
        });
      }
      resolve(versionInfo);
    } catch (e) {
      console.warn(e);
      reject(e);
    }
  });
};
