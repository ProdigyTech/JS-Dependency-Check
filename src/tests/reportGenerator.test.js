import path from "path";
import appRoot from "app-root-path";
import { existsSync, unlink } from "fs";
import {
  generateJSONReportFromRawData,
  generateHTMLReportFromRawData,
  writeReport,
} from "../reportGenerator";
import { checkDependencies } from "../checkDependencies";

const mockPackageJsonObject = {
  repoInfo: { name: "@prodigytech/js-dependency-check", version: "4.0.0" },
  dependencies: [
    { package: "@babel/preset-env", version: "^7.16.11" },
    { package: "app-root-path", version: "^3.0.0" },
    { package: "axios", version: "^0.27.1" },
    { package: "babel-jest", version: "^28.0.1" },
    { package: "jest", version: "^28.0.1" },
    { package: "semver", version: "7.3.7" },
  ],
  peerDependencies: [],
  devDependencies: [
    { package: "@semantic-release/changelog", version: "^6.0.1" },
    { package: "@semantic-release/exec", version: "^6.0.3" },
    { package: "@semantic-release/git", version: "^10.0.1" },
    { package: "rollup-plugin-terser", version: "^7.0.2" },
    { package: "semantic-release", version: "^19.0.2" },
  ],
};
const DIR_BASE = path.resolve(appRoot.path);

test("it should generate a JSON report", async () => {
  const { repoInfo, peerDependencies, devDependencies, dependencies } =
    mockPackageJsonObject;

  const rawData = await checkDependencies({
    peerDependencies,
    dependencies,
    devDependencies,
  });

  const jsonReport = generateJSONReportFromRawData(
    { ...rawData, disableTime: true },
    repoInfo
  );

  expect(jsonReport).toBeDefined();
  expect(jsonReport).toMatchSnapshot();
});

test.skip("it should generate a HTML report", async () => {
  const { repoInfo, peerDependencies, devDependencies, dependencies } =
    mockPackageJsonObject;

  const rawData = await checkDependencies({
    peerDependencies,
    dependencies,
    devDependencies,
  });

  const htmlReport = generateHTMLReportFromRawData(
    { ...rawData, disableTime: true },
    repoInfo
  );

  expect(htmlReport).toBeDefined();
  expect(htmlReport).toMatchSnapshot();
});

test("Should write the HTML report to disk", async () => {
  const { repoInfo, peerDependencies, devDependencies, dependencies } =
    mockPackageJsonObject;

  const rawData = await checkDependencies({
    peerDependencies,
    dependencies,
    devDependencies,
  });

  const htmlReport = generateHTMLReportFromRawData(
    { ...rawData, disableTime: true },
    repoInfo
  );

  await writeReport(htmlReport, "html", DIR_BASE);

  const expectedFilePath = path.resolve(
    DIR_BASE,
    "dependency-status-report.html"
  );
  expect(existsSync(expectedFilePath)).toBe(true);

  unlink(expectedFilePath, (err) => {
    if (err) {
      throw new Error(err);
    }
  });
});

test("Should write the JSON report to disk", async () => {
  const { repoInfo, peerDependencies, devDependencies, dependencies } =
    mockPackageJsonObject;

  const rawData = await checkDependencies({
    peerDependencies,
    dependencies,
    devDependencies,
  });

  const jsonReport = generateHTMLReportFromRawData(
    { ...rawData, disableTime: true },
    repoInfo
  );

  await writeReport(jsonReport, "json");

  const expectedFilePath = path.resolve(
    DIR_BASE,
    "dependency-status-report.json"
  );
  expect(existsSync(expectedFilePath)).toBe(true);

  unlink(expectedFilePath, (err) => {
    if (err) {
      throw new Error(err);
    }
  });
});

test("Write file should throw an error if data is not defined", async () => {
  try {
    await writeReport(null, "json");
  } catch (e) {
    expect(e.code).toBe("ERR_INVALID_ARG_TYPE");
  }
});
