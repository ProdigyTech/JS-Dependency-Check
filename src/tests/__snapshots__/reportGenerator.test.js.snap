// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should generate a HTML report 1`] = `
"
        <html>
        <title> Dependency Check -- Report </title>
        <head>
        </head>
        <style>
        {
        font-family: Arial, Helvetica, sans-serif;
        border-collapse: collapse;
        width: 100%;
        }

        div.wrapper{
        display: flex;
        flex-direction: column;
        flex-wrap: wrap;
        width:100%;
        }

        .header, .sub-header{
          flex-direction: row;
          width:90%;
        }
        
        .legend-table {
          flex-direction: row;
          width: 50%;
          margin-right: 3em;
        }
        
        .dep-table, .dev-table, .peer-table, .error-table {
             width: 80%;
             flex-direction: row;
             min-width: 60%;
        }
        .status {
          min-width: 5em;
        }

         td, th {
        border: 1px solid #ddd;
        padding: 8px;
        }
        tr {
            cursor: pointer;
        }
        tr:nth-child(even){background-color: #f2f2f2;}

        tr:hover {background-color: #ddd;}

        #th {
        padding-top: 12px;
        padding-bottom: 12px;
        text-align: left;
        background-color: #04AA6D;
        color: white;
        }
        </style>
    
        <body>
        <div class=\\"wrapper\\">
        <div class=header>
        <h2>Dependency Check Results for @prodigytech/js-dependency-check v4.0.0 </h2>
        </div>
        <div class=\\"sub-header\\">
          <h3>⚠️ There are  5 packages that need to be updated </h3>
        </div>
        
        <div class=\\"legend-table\\"> 
              
                    <h4>Legend </h4>
                <table id=\\"legend\\">
                    <thead>
                        <tr>
                        <td>Color</td>
                        <td>Meaning</td>
                        <td>Upgrade Type</td>
                    </thead>
                    <tbody>
                       <tr>
                           <td style=\\"background-color:green\\"></td>
                           <td>Up to date, no action needed.</td>
                           <td>UP TO DATE N/A</td>
                           </tr><tr>
                           <td style=\\"background-color:yellowgreen\\"></td>
                           <td>Patch upgrade, no breaking changes</td>
                           <td>PATCH</td>
                           </tr><tr>
                           <td style=\\"background-color:yellow\\"></td>
                           <td>Minor upgrade, possible breaking changes. 
 Consult the change log</td>
                           <td>PREPATCH PREMINOR MINOR</td>
                           </tr><tr>
                           <td style=\\"background-color:red\\"></td>
                           <td>Major upgrade with breaking changes. 
 Consult the change log</td>
                           <td>PRERELEASE MAJOR PREMAJOR</td>
                           </tr>
                    </tbody>
                    </table>
        </div>
        <div class=\\"stats-table\\">
              
                    <h4>Stats </h4>
                <table id=\\"stats\\">
                    <thead>
                        <tr>
                        <td>Upgrade Type</td>
                        <td>Package Count</td>
                    </thead>
                    <tbody>
                    
                      <tr>
                      <td>PATCH</td>
                      <td>2</td>
                      </tr>
                      
                      <tr>
                      <td>MINOR</td>
                      <td>3</td>
                      </tr>
                      
                    </tbody>
                    </table>
        </div>
        <div class=\\"dep-table\\">
                
                    <h2>Dependencies</h2>
                <table id=\\"result-table-Dependencies style=\\"width:100%\\">
                    <thead>
                        <tr>
                        <td>Package</td>
                        <td>Current Version</td>
                        <td>Current Release Date</td>
                        <td>Latest Version</td>
                        <td>Latest Version Release Date</td>
                        <td>Status</td>
                        <td> Upgrade Type </td>
                        <td> Link to package in registry </td>
                        <td> Link to package on NPM </td>
                        </tr>
                    </thead>
                    <tbody>
                       <tr>
                        <td>@babel/preset-env</td>
                        <td> 7.16.11 </td>
                        <td>1/20/2022</td>
                        <td>7.18.2 </td>
                         <td>5/25/2022</td>
                        <td class=\\"status\\" style=background-color:yellow>OUTDATED</td>
                        <td> MINOR </td>
                        <td> <a href=https://registry.npmjs.org/@babel/preset-env target=\\"_blank\\"> https://registry.npmjs.org/@babel/preset-env </a> </td>
                        <td> <a href=https://www.npmjs.com/package/@babel/preset-env target=\\"_blank\\"> https://www.npmjs.com/package/@babel/preset-env </a> </td>
                        </tr><tr>
                        <td>app-root-path</td>
                        <td> 3.0.0 </td>
                        <td>10/31/2019</td>
                        <td>3.0.0 </td>
                         <td>10/31/2019</td>
                        <td class=\\"status\\" style=background-color:green>UP TO DATE</td>
                        <td> N/A </td>
                        <td> <a href=https://registry.npmjs.org/app-root-path target=\\"_blank\\"> https://registry.npmjs.org/app-root-path </a> </td>
                        <td> <a href=https://www.npmjs.com/package/app-root-path target=\\"_blank\\"> https://www.npmjs.com/package/app-root-path </a> </td>
                        </tr><tr>
                        <td>axios</td>
                        <td> 0.27.1 </td>
                        <td>4/26/2022</td>
                        <td>0.27.2 </td>
                         <td>4/27/2022</td>
                        <td class=\\"status\\" style=background-color:yellowgreen>OUTDATED</td>
                        <td> PATCH </td>
                        <td> <a href=https://registry.npmjs.org/axios target=\\"_blank\\"> https://registry.npmjs.org/axios </a> </td>
                        <td> <a href=https://www.npmjs.com/package/axios target=\\"_blank\\"> https://www.npmjs.com/package/axios </a> </td>
                        </tr><tr>
                        <td>babel-jest</td>
                        <td> 28.0.1 </td>
                        <td>4/26/2022</td>
                        <td>28.1.1 </td>
                         <td>6/6/2022</td>
                        <td class=\\"status\\" style=background-color:yellow>OUTDATED</td>
                        <td> MINOR </td>
                        <td> <a href=https://registry.npmjs.org/babel-jest target=\\"_blank\\"> https://registry.npmjs.org/babel-jest </a> </td>
                        <td> <a href=https://www.npmjs.com/package/babel-jest target=\\"_blank\\"> https://www.npmjs.com/package/babel-jest </a> </td>
                        </tr><tr>
                        <td>jest</td>
                        <td> 28.0.1 </td>
                        <td>4/26/2022</td>
                        <td>28.1.1 </td>
                         <td>6/6/2022</td>
                        <td class=\\"status\\" style=background-color:yellow>OUTDATED</td>
                        <td> MINOR </td>
                        <td> <a href=https://registry.npmjs.org/jest target=\\"_blank\\"> https://registry.npmjs.org/jest </a> </td>
                        <td> <a href=https://www.npmjs.com/package/jest target=\\"_blank\\"> https://www.npmjs.com/package/jest </a> </td>
                        </tr><tr>
                        <td>semver</td>
                        <td> 7.3.7 </td>
                        <td>4/12/2022</td>
                        <td>7.3.7 </td>
                         <td>4/12/2022</td>
                        <td class=\\"status\\" style=background-color:green>UP TO DATE</td>
                        <td> N/A </td>
                        <td> <a href=https://registry.npmjs.org/semver target=\\"_blank\\"> https://registry.npmjs.org/semver </a> </td>
                        <td> <a href=https://www.npmjs.com/package/semver target=\\"_blank\\"> https://www.npmjs.com/package/semver </a> </td>
                        </tr>
                    </tbody>
                    </table>
        </div>
        <div class=\\"dev-table\\">
                
                    <h2>Dev Dependencies</h2>
                <table id=\\"result-table-Dev Dependencies style=\\"width:100%\\">
                    <thead>
                        <tr>
                        <td>Package</td>
                        <td>Current Version</td>
                        <td>Current Release Date</td>
                        <td>Latest Version</td>
                        <td>Latest Version Release Date</td>
                        <td>Status</td>
                        <td> Upgrade Type </td>
                        <td> Link to package in registry </td>
                        <td> Link to package on NPM </td>
                        </tr>
                    </thead>
                    <tbody>
                       <tr>
                        <td>@semantic-release/changelog</td>
                        <td> 6.0.1 </td>
                        <td>10/29/2021</td>
                        <td>6.0.1 </td>
                         <td>10/29/2021</td>
                        <td class=\\"status\\" style=background-color:green>UP TO DATE</td>
                        <td> N/A </td>
                        <td> <a href=https://registry.npmjs.org/@semantic-release/changelog target=\\"_blank\\"> https://registry.npmjs.org/@semantic-release/changelog </a> </td>
                        <td> <a href=https://www.npmjs.com/package/@semantic-release/changelog target=\\"_blank\\"> https://www.npmjs.com/package/@semantic-release/changelog </a> </td>
                        </tr><tr>
                        <td>@semantic-release/exec</td>
                        <td> 6.0.3 </td>
                        <td>12/15/2021</td>
                        <td>6.0.3 </td>
                         <td>12/15/2021</td>
                        <td class=\\"status\\" style=background-color:green>UP TO DATE</td>
                        <td> N/A </td>
                        <td> <a href=https://registry.npmjs.org/@semantic-release/exec target=\\"_blank\\"> https://registry.npmjs.org/@semantic-release/exec </a> </td>
                        <td> <a href=https://www.npmjs.com/package/@semantic-release/exec target=\\"_blank\\"> https://www.npmjs.com/package/@semantic-release/exec </a> </td>
                        </tr><tr>
                        <td>@semantic-release/git</td>
                        <td> 10.0.1 </td>
                        <td>10/29/2021</td>
                        <td>10.0.1 </td>
                         <td>10/29/2021</td>
                        <td class=\\"status\\" style=background-color:green>UP TO DATE</td>
                        <td> N/A </td>
                        <td> <a href=https://registry.npmjs.org/@semantic-release/git target=\\"_blank\\"> https://registry.npmjs.org/@semantic-release/git </a> </td>
                        <td> <a href=https://www.npmjs.com/package/@semantic-release/git target=\\"_blank\\"> https://www.npmjs.com/package/@semantic-release/git </a> </td>
                        </tr><tr>
                        <td>rollup-plugin-terser</td>
                        <td> 7.0.2 </td>
                        <td>9/4/2020</td>
                        <td>7.0.2 </td>
                         <td>9/4/2020</td>
                        <td class=\\"status\\" style=background-color:green>UP TO DATE</td>
                        <td> N/A </td>
                        <td> <a href=https://registry.npmjs.org/rollup-plugin-terser target=\\"_blank\\"> https://registry.npmjs.org/rollup-plugin-terser </a> </td>
                        <td> <a href=https://www.npmjs.com/package/rollup-plugin-terser target=\\"_blank\\"> https://www.npmjs.com/package/rollup-plugin-terser </a> </td>
                        </tr><tr>
                        <td>semantic-release</td>
                        <td> 19.0.2 </td>
                        <td>1/17/2022</td>
                        <td>19.0.3 </td>
                         <td>6/8/2022</td>
                        <td class=\\"status\\" style=background-color:yellowgreen>OUTDATED</td>
                        <td> PATCH </td>
                        <td> <a href=https://registry.npmjs.org/semantic-release target=\\"_blank\\"> https://registry.npmjs.org/semantic-release </a> </td>
                        <td> <a href=https://www.npmjs.com/package/semantic-release target=\\"_blank\\"> https://www.npmjs.com/package/semantic-release </a> </td>
                        </tr>
                    </tbody>
                    </table>
        </div>
        <div class=\\"peer-table\\">
                
        </div>
        <div class=\\"error-table\\">
            
        </div>
        </div>
        </body>
        </html>
    "
`;

exports[`it should generate a JSON report 1`] = `
"{
  \\"repoInfo\\": {
    \\"name\\": \\"@prodigytech/js-dependency-check\\",
    \\"version\\": \\"4.0.0\\"
  },
  \\"packages\\": {
    \\"devDependencies\\": [
      {
        \\"package\\": {
          \\"name\\": \\"@semantic-release/changelog\\",
          \\"registry_url\\": \\"https://registry.npmjs.org/@semantic-release/changelog\\",
          \\"npm_url\\": \\"https://www.npmjs.com/package/@semantic-release/changelog\\",
          \\"latest\\": {
            \\"version\\": \\"6.0.1\\",
            \\"releaseDate\\": \\"2021-10-30T05:33:19.426Z\\"
          },
          \\"current\\": {
            \\"version\\": \\"6.0.1\\",
            \\"releaseDate\\": \\"2021-10-30T05:33:19.426Z\\"
          },
          \\"upgradeType\\": \\"N/A\\",
          \\"error\\": false
        }
      },
      {
        \\"package\\": {
          \\"name\\": \\"@semantic-release/exec\\",
          \\"registry_url\\": \\"https://registry.npmjs.org/@semantic-release/exec\\",
          \\"npm_url\\": \\"https://www.npmjs.com/package/@semantic-release/exec\\",
          \\"latest\\": {
            \\"version\\": \\"6.0.3\\",
            \\"releaseDate\\": \\"2021-12-15T19:12:43.779Z\\"
          },
          \\"current\\": {
            \\"version\\": \\"6.0.3\\",
            \\"releaseDate\\": \\"2021-12-15T19:12:43.779Z\\"
          },
          \\"upgradeType\\": \\"N/A\\",
          \\"error\\": false
        }
      },
      {
        \\"package\\": {
          \\"name\\": \\"@semantic-release/git\\",
          \\"registry_url\\": \\"https://registry.npmjs.org/@semantic-release/git\\",
          \\"npm_url\\": \\"https://www.npmjs.com/package/@semantic-release/git\\",
          \\"latest\\": {
            \\"version\\": \\"10.0.1\\",
            \\"releaseDate\\": \\"2021-10-30T05:27:04.269Z\\"
          },
          \\"current\\": {
            \\"version\\": \\"10.0.1\\",
            \\"releaseDate\\": \\"2021-10-30T05:27:04.269Z\\"
          },
          \\"upgradeType\\": \\"N/A\\",
          \\"error\\": false
        }
      },
      {
        \\"package\\": {
          \\"name\\": \\"rollup-plugin-terser\\",
          \\"registry_url\\": \\"https://registry.npmjs.org/rollup-plugin-terser\\",
          \\"npm_url\\": \\"https://www.npmjs.com/package/rollup-plugin-terser\\",
          \\"latest\\": {
            \\"version\\": \\"7.0.2\\",
            \\"releaseDate\\": \\"2020-09-04T18:39:17.900Z\\"
          },
          \\"current\\": {
            \\"version\\": \\"7.0.2\\",
            \\"releaseDate\\": \\"2020-09-04T18:39:17.900Z\\"
          },
          \\"upgradeType\\": \\"N/A\\",
          \\"error\\": false
        }
      },
      {
        \\"package\\": {
          \\"name\\": \\"semantic-release\\",
          \\"registry_url\\": \\"https://registry.npmjs.org/semantic-release\\",
          \\"npm_url\\": \\"https://www.npmjs.com/package/semantic-release\\",
          \\"latest\\": {
            \\"version\\": \\"19.0.3\\",
            \\"releaseDate\\": \\"2022-06-09T02:10:38.012Z\\"
          },
          \\"current\\": {
            \\"version\\": \\"19.0.2\\",
            \\"releaseDate\\": \\"2022-01-18T05:27:14.724Z\\"
          },
          \\"upgradeType\\": \\"PATCH\\",
          \\"error\\": false
        }
      }
    ],
    \\"peerDependencies\\": [],
    \\"dependencies\\": [
      {
        \\"package\\": {
          \\"name\\": \\"@babel/preset-env\\",
          \\"registry_url\\": \\"https://registry.npmjs.org/@babel/preset-env\\",
          \\"npm_url\\": \\"https://www.npmjs.com/package/@babel/preset-env\\",
          \\"latest\\": {
            \\"version\\": \\"7.18.6\\",
            \\"releaseDate\\": \\"2022-06-27T19:50:53.882Z\\"
          },
          \\"current\\": {
            \\"version\\": \\"7.16.11\\",
            \\"releaseDate\\": \\"2022-01-20T15:07:51.834Z\\"
          },
          \\"upgradeType\\": \\"MINOR\\",
          \\"error\\": false
        }
      },
      {
        \\"package\\": {
          \\"name\\": \\"app-root-path\\",
          \\"registry_url\\": \\"https://registry.npmjs.org/app-root-path\\",
          \\"npm_url\\": \\"https://www.npmjs.com/package/app-root-path\\",
          \\"latest\\": {
            \\"version\\": \\"3.0.0\\",
            \\"releaseDate\\": \\"2019-10-31T15:32:11.882Z\\"
          },
          \\"current\\": {
            \\"version\\": \\"3.0.0\\",
            \\"releaseDate\\": \\"2019-10-31T15:32:11.882Z\\"
          },
          \\"upgradeType\\": \\"N/A\\",
          \\"error\\": false
        }
      },
      {
        \\"package\\": {
          \\"name\\": \\"axios\\",
          \\"registry_url\\": \\"https://registry.npmjs.org/axios\\",
          \\"npm_url\\": \\"https://www.npmjs.com/package/axios\\",
          \\"latest\\": {
            \\"version\\": \\"0.27.2\\",
            \\"releaseDate\\": \\"2022-04-27T10:00:58.685Z\\"
          },
          \\"current\\": {
            \\"version\\": \\"0.27.1\\",
            \\"releaseDate\\": \\"2022-04-26T07:36:38.268Z\\"
          },
          \\"upgradeType\\": \\"PATCH\\",
          \\"error\\": false
        }
      },
      {
        \\"package\\": {
          \\"name\\": \\"babel-jest\\",
          \\"registry_url\\": \\"https://registry.npmjs.org/babel-jest\\",
          \\"npm_url\\": \\"https://www.npmjs.com/package/babel-jest\\",
          \\"latest\\": {
            \\"version\\": \\"28.1.3\\",
            \\"releaseDate\\": \\"2022-07-13T14:12:33.491Z\\"
          },
          \\"current\\": {
            \\"version\\": \\"28.0.1\\",
            \\"releaseDate\\": \\"2022-04-26T10:02:43.209Z\\"
          },
          \\"upgradeType\\": \\"MINOR\\",
          \\"error\\": false
        }
      },
      {
        \\"package\\": {
          \\"name\\": \\"jest\\",
          \\"registry_url\\": \\"https://registry.npmjs.org/jest\\",
          \\"npm_url\\": \\"https://www.npmjs.com/package/jest\\",
          \\"latest\\": {
            \\"version\\": \\"28.1.3\\",
            \\"releaseDate\\": \\"2022-07-13T14:12:47.527Z\\"
          },
          \\"current\\": {
            \\"version\\": \\"28.0.1\\",
            \\"releaseDate\\": \\"2022-04-26T10:02:55.925Z\\"
          },
          \\"upgradeType\\": \\"MINOR\\",
          \\"error\\": false
        }
      },
      {
        \\"package\\": {
          \\"name\\": \\"semver\\",
          \\"registry_url\\": \\"https://registry.npmjs.org/semver\\",
          \\"npm_url\\": \\"https://www.npmjs.com/package/semver\\",
          \\"latest\\": {
            \\"version\\": \\"7.3.7\\",
            \\"releaseDate\\": \\"2022-04-12T17:26:24.970Z\\"
          },
          \\"current\\": {
            \\"version\\": \\"7.3.7\\",
            \\"releaseDate\\": \\"2022-04-12T17:26:24.970Z\\"
          },
          \\"upgradeType\\": \\"N/A\\",
          \\"error\\": false
        }
      }
    ],
    \\"failedLookups\\": []
  },
  \\"stats\\": {
    \\"N/A\\": 6,
    \\"PATCH\\": 2,
    \\"MINOR\\": 3
  },
  \\"reportGeneratedAt\\": {
    \\"date\\": false,
    \\"time\\": false
  }
}"
`;
